Bring back panic info to abort

danbugs opened the pull request:
In #474, I removed the guest panic context region. With that, aborting was split into two steps:
    (1) printing panic info w/ debug_print, and
    (2) exiting w/ abort and codes.
    
This PR addresses feedback I got in #474 to join the operations by buffering string output when aborting.
    
This PR also cleans up our manual chunking logic w/ outb and leverages out8, out16, and out32 to minimize VM exits. Abort detects it should finishing buffering by a terminator code (0xFF).
    
Edited tests to check for panic message in aborts as they were prior to #474 merged.

This PR can be reviewed commit by commit. I also made some other minor changes like:
- removing some legacy unused structs, and
-  adding a feature flag to logging some of our build metadata, which uses a git2 dependency that is not available on every target.

ludfjig commented on src/hyperlight_guest/src/lib.rs:
```diff
@@ -73,9 +70,11 @@ pub(crate) static _fltused: i32 = 0;
-    outb(OutBAction::Abort as u16, &[ErrorCode::UnknownError as u8]);
-    unsafe { unreachable_unchecked() }
+    let msg = info.to_string();
+    let c_string = alloc::ffi::CString::new(msg)
```
Is there a reason we're using **C** string here over regular string?

danbugs replied:
We used a C string before too. I think it's for compatibility w/ our C-API: `pub extern "C" fn hl_abort_with_code_and_message(err: i32, message: *const c_char)`.

syntactically replied:
Hm, I think there is probably a decent argument for using proper strings with their length passed properly in all our core functions and only converting to C strings at the C FFI boundaries.

danbugs replied:
Fair play. Because this way of passing a message is already in main, for now, I'll keep this as is and maybe target it in a follow-up 👍

ludfjig commented on src/hyperlight_host/src/sandbox/outb.rs:
```diff
@@ -93,6 +94,58 @@ pub(super) fn outb_log(mgr: &mut SandboxMemoryManager<HostSharedMemory>) -> Resu
+                let guest_error = ErrorCode::from(guest_error_code as u64);
+
+                match guest_error {
+                    ErrorCode::StackOverflow => {
```
What do you think about making stackoverflow it's own abort?

danbugs replied:
It's simpler that the other aborts, but I don't really see a reason to separate it. This sets precedence to custom handling of different error codes, which I think should be okay to do without creating a new OutBAction.

simongdavies commented on src/hyperlight_host/src/sandbox/uninitialized.rs:
```diff
@@ -847,7 +837,19 @@ mod tests {
     // from their workstation to be successful without needed to know about test interdependencies
     // this test will be run explicitly as a part of the CI pipeline
     #[ignore]
+    #[cfg(feature = "build-metadata")]
```
Don't understand why this is behind the build metadata feature? its testing logging and tracing?

danbugs replied:
This is only used by tests that are gated behind the build-metadata feature, so, without gating this, we get errors about dead_code.

ludfjig reviewed (commented):
Looks reasonable but have some questions. Also I'm not sure thread_local is right for this. What happens when two vms are running on the same thread?

danbugs commented on src/hyperlight_guest/src/entrypoint.rs:
```diff
@@ -43,11 +43,12 @@ pub fn halt() {
 #[no_mangle]
 pub extern "C" fn abort() -> ! {
-    abort_with_code(&[0])
+    abort_with_code(&[0, 0xFF])
```
That's right. We shouldn't require the caller to think about it. I can remove it from the `abort` fxn, but this is just a shortcut to not require 2 VM exits.

ludfjig replied:
If the 0xFF is unconditionally added by `abort_with_code`, then wouldn't adding it here add an additional exit, not reduce?

danbugs replied:
Responded w/ a top-level msg.

danbugs commented on src/hyperlight_guest/src/entrypoint.rs:
```diff
@@ -56,13 +57,19 @@ pub fn abort_with_code(code: &[u8]) -> ! {
+    let message_bytes = CStr::from_ptr(message_ptr).to_bytes(); // excludes null terminator
+
+    // Step 3: Send the message itself in chunks
+    outb(OutBAction::Abort as u16, message_bytes);
```
0xFF is outside the printable range, so a message should not contain that. I'm in favour of keeping it as one OutBAction to avoid the extra error handling we'd have to do to express the relationship between the separate actions in code...

danbugs commented on src/hyperlight_guest/src/entrypoint.rs:
```diff
@@ -56,13 +57,19 @@ pub fn abort_with_code(code: &[u8]) -> ! {
 /// # Safety
 /// This function is unsafe because it dereferences a raw pointer.
 pub unsafe fn abort_with_code_and_message(code: &[u8], message_ptr: *const c_char) -> ! {
```
We do use it—e.g., our interrupt handler sends 2 codes: one error code + one exception code.

ludfjig commented on src/hyperlight_host/src/sandbox/outb.rs:
```diff
@@ -93,6 +94,58 @@ pub(super) fn outb_log(mgr: &mut SandboxMemoryManager<HostSharedMemory>) -> Resu
+const ABORT_TERMINATOR: u8 = 0xFF;
+
+thread_local! {
+    static GUEST_ABORT_BUFFER: RefCell<Vec<u8>> = const { RefCell::new(Vec::new()) }
```
Can we limit the size of this? Could potentially be a ddos target otherwise

danbugs replied:
Will add a max length and check for it.

danbugs commented:
> Looks reasonable but have some questions. Also I'm not sure thread_local is right for this. What happens when two vms are running on the same thread (not currenly possible but will be very soon)?

Removed thread_local buffer and moved buffer to `MemMgrWrapper` so that it is exclusive to a sandbox and not to a thread.

ludfjig reviewed (changes requested):
LGTM, but old file are added back

syntactically commented on src/hyperlight_guest/src/entrypoint.rs:
```diff
@@ -56,13 +57,19 @@ pub fn abort_with_code(code: &[u8]) -> ! {
     outb(OutBAction::Abort as u16, code);
+
+    // Step 2: Convert the C string to bytes
+    let message_bytes = CStr::from_ptr(message_ptr).to_bytes(); // excludes null terminator
```
Why are we doing this? We can define the ABI for hostcalls here, since this isn't "real" port I/O and we know outb will lead to a synchronous exit.

I always assumed this would be something like
- set <some well-known register, e.g. rax> to address of {length, data} FFI string
- call outb with an abort action

And the host would take care of reading the guest pointer and then reading the bytes of the string from guest memory.

danbugs replied:
Part of the requirement here is being able to panic before memory is set on the guest side, so I can't just put the string in memory. Otherwise, yeah—I would use our output stack.

danbugs replied:
W/ the way our `out` handler is rn, there's no easy way to of doing something like ` vcpu_fd.get_regs()` to get a register value from a guest. Doing that would incur a bigger change than I had planned for this PR. I have an issue tracking improving the correctness of our `out` usage (#477). Maybe I can elaborate this as a follow-up there? I can see that removing this reliance on `out` would make it easier to transition to some other doorbell mechanism in the future.

danbugs replied:
Also, I should add—before, we already used the value part of `outb` to get data: https://github.com/hyperlight-dev/hyperlight/blob/ae139b02117ff3451a3e0a6745061ce9b4db6c34/src/hyperlight_host/src/sandbox/outb.rs#L139. In here and in #474, we're just getting more bytes.

syntactically replied:
I remember adding a similar interface for the trace stuff I did a bit ago (https://github.com/hyperlight-dev/hyperlight/pull/103/commits/8a23d78b456d7640be0b9a0f4384646e84a10753#diff-ff9dc5a5e903b284efc74dbfad22d47789d9b83e293fd82d4ab6ee3d2fc34c84, https://github.com/hyperlight-dev/hyperlight/pull/103/commits/0500ce85baf951509c01cf7aead26582673e709a), and it was pretty simple IIRC---basically just pass in the hypervisor and make sure you can get regs from it (possibly the debugging changes have already done the latter?). As @ludfjig pointed out, having this semi-state-machine that requires a bunch of different out calls in order makes me rather nervous, so I would rather avoid introducing more code that does this pattern here.

ludfjig replied:
I agree, I think it would be good to simplify this

danbugs replied:
I agree on this being the way forward. I'll need the outb handler to be able to extract register values in about 2 PRs anyway when I move the memory setup to the guest. I just feel like this would just be a hassle to do without #312 . How about we merge this to bring context back to aborts and implement your request as a follow-up? Hoping to do it for v0.5.0 still.

simongdavies replied:
I agree lets merge this add an issue for the improvements that Lucy suggested and get it done when you make the other changes, lets make sure if gets in v0.5.0

danbugs replied:
Added info to this issue: #477 .

syntactically commented on src/hyperlight_guest/src/host_function_call.rs:
```diff
@@ -119,11 +114,25 @@ pub fn outb(port: u16, data: &[u8]) {
+/// This function should be used in debug mode only. This function does not
+/// require memory to be setup to be used.
+pub fn debug_print(msg: &str) {
+    outb(OutBAction::DebugPrint as u16, msg.as_bytes());
```
Same comment as on abort above - why are we relying on the semantics of the legacy x86 port I/O instruction this way, when we really are just using it as a fast way to cause a synchronous vmexit after which the host can inspect the guest state howsoever we choose?

danbugs replied:
Like you said, currently, we use `out` to perform exits. `out` enables us to transmit a port and up to 32 bits of value. This is just another usage of the mechanism we already rely on. Here, we are transmitting a message through multiple VM exits with characters being values. It's a slow but sometimes useful mechanism to print something before memory is set. It was previously introduced in #474 as a way to print abort messages. This PR removes its usage there in favor of still passing panic info with the HyperlightError but I'm leaving it here because I think it can still be useful when debugging.

syntactically replied:
> Like you said, currently, we use out to perform exits. out enables us to transmit a port and up to 32 bits of value. This is just another usage of the mechanism we already rely on. Here, we are transmitting a message through multiple VM exits with characters being values. It's a slow but sometimes useful mechanism to print something before memory is set. It was previously introduced in https://github.com/hyperlight-dev/hyperlight/pull/474 as a way to print abort messages. This PR removes its usage there in favor of still passing panic info with the HyperlightError but I'm leaving it here because I think it can still be useful when debugging.

`out` the way we use it is a way of forcing a guest exit to VMM code, nothing more. _If we were a real hardware device on an actual port I/O bus_, yes, it would transmit a word of value to us, but we are not---the host can inspect the state of the guest arbitrarily upon exit.

danbugs replied:
Responded in the thread above.

syntactically commented on src/hyperlight_guest/src/interrupt_handlers.rs:
```diff
@@ -31,10 +31,9 @@ pub extern "sysv64" fn hl_exception_handler(
 ) {
     let exception = Exception::try_from(exception_number as u8).expect("Invalid exception number");
     let msg = format!(
-        "EXCEPTION: {:#?}\n\
```
Why does this remove useful logging?

syntactically commented on src/hyperlight_host/src/sandbox/mem_mgr.rs:
```diff
@@ -29,27 +29,39 @@ pub type StackCookie = [u8; STACK_COOKIE_LEN];
+pub(crate) struct MemMgrWrapper<S> {
+    mgr: SandboxMemoryManager<S>,
+    stack_cookie: StackCookie,
+    abort_buffer: Vec<u8>,
```
Is there a reason to have this buffer be part of this wrapper, rather than just have the abort message be returned as an Error variant and hopefully be bubbled up to the host code that initially invoked a guest function?

danbugs replied:
The buffer holds the entire panic context info. In #474, I had changed this logic to essentially just use error variants but one of the requirements here was to not lose that info.

syntactically replied:
Hm, is this an artifact of the fact that the message is being split across multiple exits?

danbugs replied:
No. Before, we also put the entire panic context info into the guest panic context memory region.

syntactically reviewed (commented):
I didn't take a super detailed look at this yet, but I'm somewhat confused by the approach.

simongdavies reviewed (approved):
I agree this can be improved, but its on the unhappy path, lets add an issue for the improvements that Lucy suggested and get this done after you make the other changes, lets make sure if gets in v0.5.0 and mark the issue as such
